command/agent.go

```diff
@@ -558,11 +570,34 @@ func (c *AgentCommand) Run(args []string) int {
 			Namespace:     namespace,
 			ExitAfterAuth: exitAfterAuth,
 		})
-		tsDoneCh = ts.DoneCh
 
-		go ah.Run(ctx, method, errCh)
-		go ss.Run(ctx, ah.OutputCh, sinks, errCh)
-		go ts.Run(ctx, ah.TemplateTokenCh, config.Templates, errCh)
+		g.Add(func() error {
+			return ah.Run(ctx, method)
+		}, func(error) {
+			cancelFunc()
+		})
+
+		g.Add(func() error {
+			err := ss.Run(ctx, ah.OutputCh, sinks)
+			c.logger.Info("sinks finished, exiting")
+
+			// Wait until templates are rendered
+			if len(config.Templates) > 0 {
```

Comment:
Should we do && ts.exitAfterAuth?

Reply:
We shouldn't need to in this case. We are only concerned about it being blocked by this https://github.com/hashicorp/vault/pull/9761/files#diff-0295527e0c39f993dedbf055a88e7b31R104.
If the there are templates to be rendered, `ts.Run` should correctly return and we would need to block here regardless of whether exit after auth is set.